kiro-synth-host/src/midi/drivers/jackmidi.rs

```diff
@@ -0,0 +1,80 @@
+use anyhow::Result;
+use thiserror::Error;
+
+use kiro_midi_core::decoder::{CallbackResult, Decoder, DecoderCallbacks};
+use kiro_midi_core::messages::Message;
+
+use crate::midi::drivers::{MidiError, MidiHandler};
+
+use jack::{Client, Control, ProcessHandler, ProcessScope};
+
+struct MidiDecoderCallbacks<'a, Handler: MidiHandler> {
+  timestamp: u64,
+  handler: &'a mut Handler,
+}
+
+impl<'a, Handler> DecoderCallbacks for MidiDecoderCallbacks<'a, Handler>
+where
+  Handler: MidiHandler,
+{
+  fn on_message(&mut self, message: Message) -> CallbackResult {
+    self.handler.on_message(self.timestamp, message);
+    CallbackResult::Continue
+  }
+
+  fn on_sysex(&mut self, data: &[u8]) -> CallbackResult {
+    self.handler.on_sysex(self.timestamp, data);
+    CallbackResult::Continue
+  }
+}
+
+#[derive(Error, Debug)]
+pub enum JackMidiError {
+  #[error("Error creating a new client")]
+  ClientCreate(#[source] jack::Error),
+
+  #[error("Error creating an input port")]
+  PortCreate(#[source] jack::Error),
+}
+
+struct JackProcessHandler<'a, Handler: MidiHandler> {
+  input_port: jack::Port<jack::MidiIn>,
+  decoder: Decoder<'a>,
+  handler: Handler,
+}
+
+impl<'a, Handler: MidiHandler> ProcessHandler for JackProcessHandler<'a, Handler> {
+  fn process(&mut self, _: &Client, ps: &ProcessScope) -> Control {
+    for raw_midi in self.input_port.iter(ps) {
+      let timestamp = raw_midi.time as u64;
+      let mut callbacks = MidiDecoderCallbacks {
+        timestamp,
+        handler: &mut self.handler,
+      };
+      let mut source = raw_midi.bytes.iter().copied();
```

Comment:
A note for me: This made me realize that I might need to re-visit the decoder interface to avoid having to copy the raw bytes.

Reply:
I tried and managed to solve it to avoid copying the data, but the changes were not that trivial. I'm thinking on doing it in master so you can rebase it.

Reply:
You can rebase now, and remove the `.copied()`.